### ISSUE TITLE
Records search feature

### ISSUE THREAD

**Issue by dpalomino** [Cadasta/cadasta-platform#825]
We need a feature to execute search queries and present records that match that query. 

Requirements are included [in the wiki](https://devwiki.corp.cadasta.org/Search%20and%20Filtering). 

Several tasks are included here: 
- Text-based queries for any text field (queries applying to any field text)
- Text-based queries for specific fields (being able to specify which fields are considered in the query)
- Number-based queries (for integer, decimal, and date/time types)
- Presentation of search results (on a map and a list of records), and accessing to the location details
- Being able to export the results in *.csv*

**Comment by seav** [comment]
I suggest that before we proceed with tackling this feature, we should answer the usual engineering question: build or buy? (Do we want to create our own custom search engine, or can we leverage an existing search engine?)

**Comment by amplifi** [comment]
I recommend using Elasticsearch; it's open source, easy to implement, solid REST API, powerful, scalable, reliable, etc. Includes all the features from our requirements and more. Even if we had time allotted to build an engine in-house, the ongoing optimization, scaling, and maintenance would be a job in itself.

**Comment by dpalomino** [comment]
Thanks a lot @seav, @amplifi for the input. Sure, if there is something open source that we can reuse and adapts well to our needs, much better. 

Taking a look to [Lucene syntax](https://lucene.apache.org/core/2_9_4/queryparsersyntax.html) ([Elasticsearch](https://en.wikipedia.org/wiki/Elasticsearch) based on it) it seems that cover our requirements regarding search in text fields, dates or integers.

**Comment by seav** [comment]
Here are some suggestions on implementing the search feature using Elasticsearch (ES) based on my (limited) research and experiments. Please feel free to comment, ask questions, and poke holes. Hopefully we can then refine this and so have a clearer idea of how to implement search for Sprint 11 (and beyond?).

### Infrastructure
ES is intended to be run as a cluster of nodes with a single node providing the API endpoint. I assume that this is what @amplifi has already provisioned in AWS. Note that the API has no authentication by default (thought there's the [Shield](https://www.elastic.co/guide/en/shield/current/getting-started.html) plugin for that), but to make things simple, I would suggest just implementing an isolated cluster and only allowing the Cadasta platform server access to the API via routing/firewall restrictions and only allowing HTTPS access to the API for added security.

This also means that we need three ES clusters, one each for the staging, demo, and production platforms if we want search to be functional on all three. But these 3 clusters can all probably share the same physical/virtual nodes—we just set different IP ports for the API.

### ES data structure
_Note: To make sense of this section, please read the [Basic Concepts](https://www.elastic.co/guide/en/elasticsearch/reference/current/_basic_concepts.html) section of the ES docs._

Based on the requirements, there is no need to have inter-project search capability. So every search request will always be within a single project. Therefore, I suggest that each project have its own individual ES _index_. And each class of records (location, party, 3 types of relationships) be implemented as an ES _type_. Having separate indexes for the projects would make search more performant since ES doesn't need to filter irrelevant records from other projects. This also allows us to fine-tune each index's performance attributes based on the amount of data in each project.

This structure implies that we can't use many Django-ES libraries or wrappers like the popular [Haystack](http://haystacksearch.org/) plugin. Many such libraries/wrappers assume that the Django website will only need a single index for search. So we would need to implement a custom solution. But we could still use helper libraries like the [Python Elasticsearch Client](http://elasticsearch-py.readthedocs.io/en/master/index.html) to simplify coding some stuff.

### Data flow
To first populate the ES indexes, I suggest adding a `reindexsearch` management command which drops everything and then reindexes the whole database. I guess we can also allow an optional project slug argument so that only that project's index gets reindexed.

To simplify lookup and update of individual records, I suggest that we reuse the records' ID as their document ID in ES too.

When a record is created/updated/deleted, I think we can just add a `post_save` signal that updates the record's document in ES. The big question that I see is whether this operation should be synchronous or not. Note that we are accessing ES through a network so we potentially have network reliability issues. I am currently leaning towards making the search index update synchronous to make things simpler, but if there's a lightweight async way (fork a sub-process or thread?) then that would be better. Any failures to update the index should be logged so that we can correct things and prevent the index from slowly becoming out of sync. There are potentially some rare race conditions possible with async (two people updating the same record at the same time and they get updated in ES in the wrong order) since async is not atomic by definition.

### Making queries
Any query language we expose in the UI needs to be translated into the [ES query language](https://www.elastic.co/guide/en/elasticsearch/reference/current/_introducing_the_query_language.html), but this should be pretty easy to implement. Pagination is also supported by ES, but because ES does not maintain state across API accesses, it is possible that data would be inconsistent when the user pages through the search results (a newly added record could appear in the results) or subsequently exports the results into an Excel file. I don't think this is a problem.

### Django innards
For Django itself, I suggest creating a `search` app to contain all search-related code. And because the documents indexed by ES are just JSON documents, we can reuse the existing DRF serializers, though I think they need to be modified to allow a more "flat" structure since that is closer to how ES actually indexes such documents (so no GeoJSON for locations) and JSON attributes should be folded into the top JSON level instead of residing under the `"attributes"` field.

### Testing
For the dev VM, basically we need to mock the ES cluster. I guess we can just implement a simple HTTP server that allows us to dynamically program the HTTP response in the tests. This server should be reusable for both unit tests and functional tests.

**Comment by amplifi** [comment]
A few notes:

Infrastructure has already been provisioned and configured for a single cluster with network-, group-, and user-level access restrictions allowing valid traffic from the platform server only.

> But these 3 clusters can all probably share the same physical/virtual nodes—we just set different IP ports for the API or designate different nodes for the endpoints.

This isn't possible under the existing platform architecture; it would violate our environment isolation requirements. It'd also expose us to a scenario where issues causing excessive processing times on our staging indices would impact our demo and production performance. We'll need to evaluate whether search is necessary for all three environments, and if so, implement our ES model for each.

> To first populate the ES indexes, I suggest adding a reindexsearch management command which drops everything and then reindexes the whole database. I guess we can also allow an optional project slug argument so that only that project's index gets reindexed.

We'll want to implement indexing off-the-bat with more granularity, especially when it comes to situations where a release version includes a significant migration. With larger data sets, this will be a non-trivial operation and running a system-wide reindexing won't be feasible once live.

This is also where making good use of aliases right from the start will be a huge help to avoid any downtime during reindexing.

> I am currently leaning towards making the search index update synchronous to make things simpler, but if there's a lightweight async way (fork a sub-process or thread?) then that would be better.

Index update processes have long execution times by nature; Lucene writes are expensive. Updates take longer than the initial indexing. For our purposes, we should seek to avoid too-frequent updates and prioritize batch updates via the bulk API (which allows ES to leverage its own optimizations to reduce run times, minimizes connection times and request overhead especially with SSL, etc). Given our data sets, we have less of a need for immediate sync than for reliable performance. We should use an asynchronous process for updates; a queue with deduplication (to prevent the rare race condition you mentioned) run via a single background worker or using transactional locking -- this also greatly simplifies error handling and prevents replication issues. Initial runs or re-indexing should use ranges executed periodically to avoid delays from queuing large volumes of data. Combining these two methods to cover the two distinct indexing cases should yield the best performance and let us stagger workload more efficiently.

> Pagination is also supported by ES, but because ES does not maintain state across API accesses, it is possible that data would be inconsistent when the user pages through the search results (a newly added record could appear in the results) or subsequently exports the results into an Excel file. I don't think this is a problem.

For pagination, results are sorted and can be accessed using the 'size' and 'from' parameters, so combining that with bulk updates as above we can basically ensure a user gets a consistent set of results per user query. We should set a conservative limit for max results up front to prevent queries from getting too intensive.

This can all be discussed further in the search call next week :)

**Comment by wonderchook** [comment]
@seav how would this work with django-tutelary and the current permissions?

**Comment by seav** [comment]
@wonderchook, I am assuming that any permissions on records will only be stored in the platform and not baked into the search index because permissions is tied to the user and is not an attribute of the records. So when records are returned as search results, we would need to check the user's permission to view each record and then discard records from the results if the user doesn't have view permissions. In addition, I assume that if a user doesn't have the list permission, then the search function would not be available to them.

This would be complicated and inefficient to implement in a user-friendly way though because we can't say upfront that there are a total of _N_ results because computing _N_ based on permissions is inefficient, and paging through the results would be complex. (You can ask Elasticsearch to return results 11–20 but what happens if the 19th result is discarded? Do you make another query to retrieve the 21st result? But what if _that_ is also discarded?)

An alternative is to display the results whether the user has permission to view the resulting records or not. But if the user follows the link to see a record's detail, then they get a permission denied error. This is not satisfactory from a privacy point of view though because the search results would necessarily leak information about the records.

**Comment by clash99** [comment]
Link to wireframes (wip): https://drive.google.com/open?id=0BzpiEtMtHC3rVFlVelgzNHBxaUk

Still working on advanced options and map (versus list) option.

**Comment by clash99** [comment]
Here is the proposed flow for a Public User for Projects and Organizations:

![2 0_ public user](https://cloud.githubusercontent.com/assets/1119621/20364522/0ebc3878-ac09-11e6-8516-e029c87e2ae3.png)

![3 0_ public user](https://cloud.githubusercontent.com/assets/1119621/20364566/40de67c2-ac09-11e6-8577-416fd1dfd534.png)

For the new messaging on the sign in page I have updated it to:

Because we care about protecting your sensitive data, you must sign in to verify you have the proper permission to view project information. Access to project data is restricted to current project members only. Please sign in to continue.

**Comment by clash99** [comment]
Here is the layout flow for Project and Organizations for the different user roles:  https://drive.google.com/a/cadasta.org/file/d/0BzpiEtMtHC3renRpeXVZLXNyQlE/view?usp=sharing

In summary here are my recommendations:
- Remove the inner subnav for non-project members for both projects and organizations
- Remove project overview stats for public users
- Add new sign in language on project overview for public users
- Add messaging to sign in page for public users
- Add option for registered users without project permissions to request project access (not sure if we are ready for this one)

Take a look and let me know what you think.  Thanks, Chandra

**Comment by clash99** [comment]
I've updated the search wireframes based on our latest discussion: https://drive.google.com/open?id=0BzpiEtMtHC3rVFlVelgzNHBxaUk

I have the phases broken out as such:

**Phase 1**
- Adding keyword search box
- New simplified search results page (no tabs, narrow results options, etc.)
- No changes to other platform pages

**Phase 2**
- Adding entity tabs
- Adding narrow results options in sidebar
- Export results feature
- Advanced search feature
- No changes to other platform pages

**Phase 3** (not included in mocks yet)
- Adding save feature
- Adding icons to entities
- Adding map/list view options
- Changes to other platform pages

**Additional Options**
- Auto-suggest feature

**Comment by dpalomino** [comment]
Thanks really a lot @clash99 for the fantastic wireframes! I think that covers everything for phase 1, and for the following. 

@seav @linzjax please let us know if you this makes sense to you.

**Comment by oliverroick** [comment]
Some things we should discuss before we start implementing the search UI.

### General

- The timeline to complete the full implementation until the end of sprint 13 is unrealistic. We are already halfway through sprint 11; I think, phase one and two will take at least two sprints each.

### Phase 1

- Sort dropdown: “Sort by relevance” — is this something we get from Elastic search?
- Pagination: We need to think about how we can do this. Currently, we have pagination only available for table views; we are using the datatables plugin to add pagination to tables. The search result can not be shown as a table; it will probably require some customized solution that we would need to build ourselves. 
- There’s a link for “search guidelines” — who is going to write them?

### Phase 2

I don’t like the split between tabs on top of the search results and “narrow results” checkboxes on the left. They should not be separated because they both address the same goal: Narrowing down the search results. 

Wouldn’t it be possible to combine both into the left-hand-side checkbox panel? Initially, you get checkboxes for locations, relationships, parties and resources. When you select a checkbox, the subcategories expand, e.g. Community Boundary and Parcel for locations.

**Comment by dpalomino** [comment]
Thanks a lot @oliverroick for the feedback, let's discuss later on during the dev call. 

Regarding the "search guidelines", I can draft the texts for this.

**Comment by dpalomino** [comment]
I've drafted some search guidelines in this doc: 
https://docs.google.com/document/d/16RevzILdt8HrWd5TEX9D5R7nQoEef1kf4x104FsUEXU/edit#

- Phase 1: only basic query build advices
- Phase 2: Phase 1, plus:
   - help for advanced queries 
   - filters (sidebar, entity tabs), downloads
   - help for the advanced search GUI
- Phase 3: TBD

**Comment by clash99** [comment]
I've updated the search wireframes with the following:

- Removed sprint numbers 
- Reworked search to exclude tabs and only be in right sidebar.  Dynamic presentation of options will be based upon users selection.  The presents more of a funnel-down approach.  The tab option is still viable if we would like to discuss as a team.

**Comment by dpalomino** [comment]
Apart from the bugfixing that could appear, I think we can move this issue to "Released" state.
